### PR TITLE
[Types] Fix entity-service's parameters inference for generic mapped types with deep literal intersections

### LINKED ISSUE #18263: Populate generated types doesn't work with filter on entityService
### Required System information

- Node.js version: v16.20.0
- NPM version: 8.19.4
- Strapi version: 4.13.7
- Database: postfgres
- Operating system: wsl
- Is your project Javascript or Typescript: Typescript

### Describe the bug

I'm using entityService with the generated types, typing my strapi, and I note when a add a filter on my api, the populate fields type are gone from my types.

Before add filters (without error):
image

On add filters (with type error):
image

### Steps to reproduce the behavior

1. Creata a collection type with relation
2. Create a api with entityService without filters and populate and check the types
3. Add filters and check the types

### Expected behavior

Don't gone my populated fields from types.

### Additional context

I don't know if is a bug or I'm using wrong, if anyone can help, thanks.

### PR DESCRIPTION
**Pull request by Convly**
### Summary
Fixed the `sort` and `filters` object notations for the entity service causing invalid inference, thus breaking the resulting types.

### Original Issue
[ #18263 - Populate generated types doesn't work with filter on entityService](https://github.com/strapi/strapi/issues/18263)

### Context
`findMany` in the entity-service is defined as (simplified version):
```ts
findMany<
    TContentTypeUID extends Common.UID.ContentType,
    TParams extends Params.Pick<
      TContentTypeUID,
      | 'filters'
      | 'sort'
      | 'populate'
	  // | etc...
    >
  >(uid: TContentTypeUID, params?: TParams): Promise<Result<TContentTypeUID, TParams>[]>
```

#### TL;DR
**When using the entity-service's `findMany` method with a `sort` object notation, it breaks the parameters type inference.
This is caused by the usage of generics (inferred upon usage) in mapped types with loose index signatures conflicting with each other.**

**Before**
```ts
export type ObjectNotation<TSchemaUID extends Common.UID.Schema> = {
  // First level sort
  [key in SingleAttribute<TSchemaUID>]?: OrderKind.Any;
} & {
  // Deep sort, only add populatable keys that have a
  // target (remove dynamic zones and other polymorphic links)
  [key in Attribute.GetKeysWithTarget<TSchemaUID>]?: ObjectNotation<
    Attribute.GetTarget<TSchemaUID, key>
  >;
};
```

#### Details
- `TParams` (`findMany`) is using a `TContentTypeUID` generic which will be inferred upon usage. (as opposed to using a hard-coded literal string like `api::restaurant.restaurant`)
- The `TContentTypeUID` generic is then used across different types in `TParams`, including inside the `sort` `ObjectNotation`
- `ObjectNotation` makes an intersection between two mapped types whose index signatures are built using the given `TSchemaUID` generic (== `TContentTypeUID` from `TParams<TContentTypeUID>`)

  > Note:  When using hard-coded literals instead of generics in the `findMany`, it'll limit the breadth of possible types and make it easier for TypeScript to infer the types down the line, which is not the case when it has to infer it from usage.
- All of this results in `ObjectNotation` creating an intersection of two mapped types whose index signatures end up being too loose and conflict with each other. It creates an errored type's intersection and breaks the inference up to `TParams`.
- With `TParams` being broken, it defaults to the abstract representation with a generic `Common.UID.Schema` being passed instead of the UID inferred from usage.
- Finally, since `TParams` is abstract, `Result` can't infer the `fields` and `populate` params anymore, and defaults to every scalar field + no populatable fields, breaking the final type and causing the initial error reported by users.

#### Proposed Solution

To solve this issue, I've decided to remove the faulty intersection, using a single mapped type and deferring the type narrowing to every value inside the map.

- This removes the conflicting index signatures. It now uses a single index built from the union of the 2 old ones.
- The type narrowing is done inside each key/value pair, allowing TypeScript to infer the correct type.
- Ultimately, this fixes the `TParams` inference, hence allowing `Result` to use the correct `fields` and `populate` params

**After**
```ts
export type ObjectNotation<TSchemaUID extends Common.UID.Schema> = {
  [TKey in ObjectNotationKeys<TSchemaUID>]?: TKey extends SingleAttribute<TSchemaUID>
    ? // First level sort (scalar attributes, id, ...)
      OrderKind.Any
    : // Deep sort (relations with a target, components, media, ...)
      ObjectNotation<Attribute.GetTarget<TSchemaUID, TKey>>;
};

type ObjectNotationKeys<TSchemaUID extends Common.UID.Schema> =
  | SingleAttribute<TSchemaUID>
  | Attribute.GetKeysWithTarget<TSchemaUID>;

```

#### TL;DR
**When using the entity-service's `findMany` method with a `filters` object notation that contains literal filters (enums), it breaks the parameters type inference.
This is caused by the usage of generics (inferred upon usage) in mapped types with incorrect inference, creating impossible situations.**

**Before**
```ts
export type ObjectNotation<TSchemaUID extends Common.UID.Schema> = {
  [TIter in Operator.Group]?: ObjectNotation<TSchemaUID>[];
} & {
  [TIter in Operator.Logical]?: ObjectNotation<TSchemaUID>;
} & ([AttributeUtils.GetScalarKeys<TSchemaUID>, AttributeUtils.GetNestedKeys<TSchemaUID>] extends [
    infer TScalarKeys extends AttributeUtils.GetScalarKeys<TSchemaUID>,
    infer TNestedKeys extends AttributeUtils.GetNestedKeys<TSchemaUID>
  ]
  ? // If both the scalar and nested keys are resolved, then create a strongly
  // typed filter object, else create a loose generic abstraction.
  Utils.Expression.If<
    Utils.Expression.And<
      Utils.Expression.IsNotNever<TScalarKeys>,
      Utils.Expression.IsNotNever<TNestedKeys>
    >,
    // Strongly typed representation of the filter object tree
    {
      [TIter in IDKey | TScalarKeys]?: AttributeCondition<TSchemaUID, TIter>;
    } & {
    [TIter in TNestedKeys]?: NestedAttributeCondition<TSchemaUID, TIter>;
  },
    // Generic representation of the filter object tree in case we don't have access to the attributes' list
    {
      [TKey in string]?:
      | AttributeCondition<TSchemaUID, never>
      | NestedAttributeCondition<TSchemaUID, never>;
    }
  >
  : never);
```

#### Details
- `TParams` (`findMany`) is using a `TContentTypeUID` generic which will be inferred upon usage. (as opposed to using a hard-coded literal string like `api::restaurant.restaurant`)
- The `TContentTypeUID` generic is then used across different types in `TParams`, including inside the `filters` `ObjectNotation`
- `ObjectNotation` infers union types (from the given Schema keys) as variables (`TScalarKeys`, `TNestedKeys`)
  ```ts
    ([AttributeUtils.GetScalarKeys<TSchemaUID>, AttributeUtils.GetNestedKeys<TSchemaUID>] extends [
        infer TScalarKeys extends AttributeUtils.GetScalarKeys<TSchemaUID>,
        infer TNestedKeys extends AttributeUtils.GetNestedKeys<TSchemaUID>
    ]
    ```
- The mapped types built using those keys as indexes have values defined as unions and intersections of types resolved from both the given Schema and `keys` (e.g. for `email` fields, it'll return `string | ({ $eq?: string } & { ... })`).
- When this happens, if the resolved type for a given generic Schema and attribute name (`key`) is a literal value (e.g. enum fields), it then creates unions/intersection of a literal with object types.
   > Note:  When using hard-coded literals instead of generics in the `findMany`, it'll limit the breadth of possible types and make it easier for TypeScript to infer the types down the line, which is not the case when it has to infer it from usage.
- Based on last points, what is likely happening is that `TScalarKeys` is inferred incorrectly upon usage, which ends up creating intersections of mapped types (based on every key), and fail to intersect literal with loose types (such as `objects` or scalars like `number`)
- With `TParams` being broken, it defaults to the abstract representation with a generic `Common.UID.Schema` being passed instead of the UID inferred from usage.
- Finally, since `TParams` is abstract, `Result` can't infer the `fields` and `populate` params anymore, and defaults to every scalar field + no populatable fields, breaking the final type and causing the initial error reported by users.

#### Proposed Solution

To solve this issue, I've simplified the code, then used a two-steps inference.

Basically, the fix is going from:

`X infer TX extends X ? ... : never`

to:

`X infer TX ? TX extends X ? ... : never : never`

Which apparently forces TypeScript to better infer types (even using generic defined upon usage).

I've tried to make some research in order to understand the difference between both, but couldn't find much. Any additional information would be greatly appreciated

**After**
```ts
export type ObjectNotation<TSchemaUID extends Common.UID.Schema> =
  RootLevelOperatorFiltering<TSchemaUID> & AttributesFiltering<TSchemaUID>;

export type AttributesFiltering<TSchemaUID extends Common.UID.Schema> = Utils.Guard.Never<
  // Combines filtering for scalar and nested attributes based on schema UID
  ScalarAttributesFiltering<TSchemaUID> & NestedAttributeFiltering<TSchemaUID>,
  // Abstract representation of the filter object tree in case we don't have access to the attributes' list
  {
    [TKey in string]?:
        | AttributeCondition<TSchemaUID, never>
        | NestedAttributeCondition<TSchemaUID, never>;
  }
>;

export type ScalarAttributesFiltering<TSchemaUID extends Common.UID.Schema> =
  AttributeUtils.GetScalarKeys<TSchemaUID> extends infer TKeys // <--|
    ? TKeys extends AttributeUtils.GetScalarKeys<TSchemaUID> //   <--| Two steps inference to force the correct narrowing 
      ? IDFiltering & { [TKey in TKeys]?: AttributeCondition<TSchemaUID, TKey> }
      : never
    : never;

export type NestedAttributeFiltering<TSchemaUID extends Common.UID.Schema> =
  AttributeUtils.GetNestedKeys<TSchemaUID> extends infer TKeys // <--|
    ? TKeys extends AttributeUtils.GetNestedKeys<TSchemaUID> //   <--| Two steps inference to force the correct narrowing
      ? { [TKey in TKeys]?: ObjectNotation<Attribute.GetTarget<TSchemaUID, TKey>> }
      : never
    : never;

export type RootLevelOperatorFiltering<TSchemaUID extends Common.UID.Schema> = {
  [TIter in Operator.Group]?: ObjectNotation<TSchemaUID>[];
} & {
  [TIter in Operator.Logical]?: ObjectNotation<TSchemaUID>;
};
```

### Why is it needed?

Issues are being reported indicating that using sort or certain filters alongside populate breaks the entity service's types.
It aims at fixing those faulty behaviors.

### How to test it?

- Go in the `getstarted` project, make sure types are generated
- Copy the `types/` folder inside the `examples/kitchensink-ts` directory
- Open the `kitchensink-ts/src/index.ts` file and heads to any method inside, make it async
- Use the following query as a base for future tests
  ```ts
  const result = await strapi.findMany('api::restaurant.restaurant', { populate: ['menu'] });
  
  console.log(result[0].menu) // this is where the error was in the issue 
  ```
- To test the `sort` fix, add a `sort` property to the params (it was failing only on object notations)
  - Whatever `sort` you're adding, `result[0].menu` must not fail anymore
- To Test the `filters` fix, add a `filters` property to the params (it was failing only on enum filters)
  - `result[0].menu` must not fail anymore. Even when using filters such as `{ priceRange: 'cheap' }`
- Feel free to play with the overall params object in order to find additional issues

### Related issue(s)/PR(s)

improve #19082
fix #18263 

---
_oh, and it also fixes the core/utils package's watch script, because why not_

### CODE REVIEW

File: packages/core/types/src/modules/entity-service/params/sort.ts [hunk 1]
```diff
@@ -75,16 +75,24 @@ export type ArrayNotation<TSchemaUID extends Common.UID.Schema> = Any<TSchemaUID
...
 
+/**
+ * Represents the keys of an object notation for a sort
+ * - SingleAttribute<TSchemaUID> represents a union of every non-populatable attribute based on the passed schema UID
+ * - Attribute.GetKeysWithTarget<TSchemaUID> provides keys with a target from the passed schema UID.
+ *
+ * This means that every member of ObjectNotationKeys can represent either a single non-populatable attribute or an attribute with a target.
+ */
+type ObjectNotationKeys<TSchemaUID extends Common.UID.Schema> =
+  | SingleAttribute<TSchemaUID>
```
Comment (alexandrebodin) [inline packages/core/types/src/modules/entity-service/params/sort.ts:93]: you don't have to change it but that type sounds exotic. Shouldn't it be in some Attribute.GetXXXXKeys util ?

File: packages/core/types/src/modules/entity-service/params/sort.ts [hunk 2]
```diff
@@ -75,16 +75,24 @@ export type ArrayNotation<TSchemaUID extends Common.UID.Schema> = Any<TSchemaUID
...
-  // Deep sort, only add populatable keys that have a
-  // target (remove dynamic zones and other polymorphic links)
-  [key in Attribute.GetKeysWithTarget<TSchemaUID>]?: ObjectNotation<
-    Attribute.GetTarget<TSchemaUID, key>
-  >;
+  [TKey in ObjectNotationKeys<TSchemaUID>]?: TKey extends SingleAttribute<TSchemaUID>
+    ? // First level sort (scalar attributes, id, ...)
+      OrderKind.Any
+    : // Deep sort (relations with a target, components, media, ...)
+      ObjectNotation<Attribute.GetTarget<TSchemaUID, TKey>>;
```
Comment (alexandrebodin) [inline packages/core/types/src/modules/entity-service/params/sort.ts:82]: Considering the key type is not explicit but coming from another type  ObjectNotationKeys, it could get changed and the conditions below would not work anymore. 

I would add a TKey extends Attribute.GetKeysWithTarget<TSchemaUID> to be exhaustive and avoid breaking if someday we change ObjectNotationKeys

File: packages/core/types/src/modules/entity-service/params/filters/index.ts [hunk 3]
```diff
@@ -9,34 +9,78 @@ export { Operator };
...
- * Filter representation for scalar attributes
+ * Definition of scalar attribute filtering for a given schema UID.
+ * @template TSchemaUID - The UID of the schema.
+ */
+export type ScalarAttributesFiltering<TSchemaUID extends Common.UID.Schema> =
+  AttributeUtils.GetScalarKeys<TSchemaUID> extends infer TKeys
+    ? TKeys extends AttributeUtils.GetScalarKeys<TSchemaUID>
+      ? IDFiltering & { [TKey in TKeys]?: AttributeCondition<TSchemaUID, TKey> }
+      : never
+    : never;
```
Comment (alexandrebodin) [inline packages/core/types/src/modules/entity-service/params/filters/index.ts:60]:
```suggestion
export type ScalarAttributesFiltering<TSchemaUID extends Common.UID.Schema> = IDFiltering &
  AttributeUtils.GetScalarKeys<TSchemaUID> extends infer TKeys
    ? TKeys extends AttributeUtils.GetScalarKeys<TSchemaUID>
      ? { [TKey in TKeys]?: AttributeCondition<TSchemaUID, TKey> }
      : never
    : never;
```

Should this be like that ? Not sure that would work will with the never 🤔

**Review by alexandrebodin — COMMENTED** [review]
Looks also much simpler to read splitted like that 👍

File: packages/core/types/src/modules/entity-service/params/filters/index.ts (hunk 3 above)
Comment (Convly) [reply packages/core/types/src/modules/entity-service/params/filters/index.ts:60]: It'll factorize to the same thing. In both examples, if it cannot infer the scalar keys it'll end up being never for the resulting type.

As for the form itself, I find that less confusing since you understand the above explanation at first glance. I'm not against yours though. Whatever feels the most elegant & readable :shrug:

File: packages/core/types/src/modules/entity-service/params/sort.ts (hunk 2 above)
Comment (Convly) [reply packages/core/types/src/modules/entity-service/params/sort.ts:82]: Yeah, I thought about that and told myself it's unlikely etc... But you're right, it's probably better to be safe. I'll add it

File: packages/core/types/src/modules/entity-service/params/sort.ts (hunk 1 above)
Comment (Convly) [reply packages/core/types/src/modules/entity-service/params/sort.ts:93]: This is only related to the sort object notation though. I don't think it makes sense to expose it elsewhere (especially from attribute utils)

File: packages/core/types/src/modules/entity-service/params/sort.ts (hunk 2 above)
Comment (Convly) [reply packages/core/types/src/modules/entity-service/params/sort.ts]: done in [64d75d0](https://github.com/strapi/strapi/pull/19093/commits/64d75d007fb872c50f0a37eed1518092ba61f425)

**Comment by Convly** [comment]
Although the fix is working in 99% of the use cases, when using an enum field (like `priceRange` in my example) alone in a filters object notation (nested or not), it leads to unwanted behavior (see image) with ObjectNotation not being inferred from usage.
image

I've spent way too much time and energy today trying to figure out where that particular issue is coming from. I'll continue tomorrow.

File: packages/core/types/src/modules/entity-service/params/filters/index.ts (hunk 3 above)
Comment (alexandrebodin) [reply packages/core/types/src/modules/entity-service/params/filters/index.ts:60]: If there is no bug no need to change I was thinking ID should be there regardless of the scalar check :)

**Comment by Convly** [comment]
Managed to make a minimal reproduction example for whoever is interested

[Playground link](https://www.typescriptlang.org/play?#code/PTAEBcE8AcFNQKoEkAioC8oAGASA3gM7gBOAlgHYDmAvgHT5FlXVYBQI7wnEM8AYqQA24WMQIAeACrI0sAB4jyAEwKJUAPgyg81TnpCgAZgFdyAY3CkA9uSMUlUmaHmKValABpQkgAoBDYj8AW1UXWGVVKwAjACtYC3UACmNSJQAubxkvaADgggzfXJCASgL-QJDtblBQYlhwY2JbHIqCAG5uXQ4DMxsiWtgCY2EtQ3tEgHI-aFJaQysrCa88aqMFjImogInO4o7WChFiQz8zeABlMwALWCC-VRWaqZm5hYmMvFByYNgMxgpKF5wsYght5otQAAfUCbbZeciDETpUAAJXiVmIDn+VC8pgA1uQrAB3ciaageVg1AykQygInwXpBILhcAQG6gQQUeBRWCCYleUisogBcCqIkYvEA0A8sx+YwEC7XW73ADaMgAuqBZeQJqyeaA-ORIOArlLeQq2YaYdNZuCJhNVAA3PyCYywSkwrbEV6LD4QQWCX6gbGArVWUzgDLkEE84igcmsXSsKBwdxaPGwSBWWmXG53dqsZO8VELcAAGVgjt5AHk4IFwBiBMJRACpLnlU4whF3JpMI9QCrJEgjqAKKAAESGpTjqETjHj9UAfgy1di8XAADkrOA-JYbG2lXcZOoVeqOkmDI6AAwrsiUCgusNKWBeKJ1Px4sU3BFVuNW4FBKOqjQHUCrkOARapgAguAJCkFExgiAQTZHK2kjtkeqDOAo4RuMeWj9oOADSmajrYGZZjmh6quh1EakuGQYTRTEasRmZnomhaXgAjBkAhyKAVwurS2ZsvApAEEMgxAtGgEujY8BEEIgjSu+eKvohziyaAADUoDbjccZjM2cbisQn7cCm8AwXBCFIShLZUAeeZ+J2OHdvhmAUaJTEEIOLGoJqXZuBQhiiN4JGQAQqyLhFmahO5bjeVRLl+bRLkaqsNSxZ8bGQGRcVRQxoC+f5dGBXlmpdGANQ1FGlaiKs9W-oWVmgKucQWFuO57uQzkdlhwWqJ5JbbhWVaCLWoi7o2QioU56UDSgmgAGSgDZZB2YMDlMJQ-WYct+wmOYvV2MoACyhqQOIHrSINiXDagFI1IUrTYa4qgdeu3W7tYfV3ctoAGPKUp+IJARKL0z5KGm8xxtEnWsoSPV-WRRCwH4MOiSa4nkGFxB1DDAOgHcGajqyZl4qwSQegA+ikyIA89oAtHkZRFNFpTeOUeRVDUdQNE0LMc+eXFgABoDiJopio6JnJHI+IaqIkY4mhJYbgS4xRi3SEqqOI6CaKBwysm16uhaIhNGMQViAazITcCcQj64bAxDCMZuqM+Jwm7AMMNtzHMa4wfiHKAiRfV126-fuzy2gs6ja9wAASVi-rUV4eMQ3FZwATB4tCFxAVjBrA8A427Js8HAOuUwQ8IlxLY729FvTkP0xBXqM9iXUakw2j6SzaN8zJgm81B7LXesySCQFfNuBp8giBqm+yNsL4GE2sG3Hfcd3F1Xf3Lx2ssAFj4s8I-OfEwT-sBhO4I9dabP6vycvqvsjgU7b30rLELn+8lC90gEfeOF9tBTgyCqPAZ9PTbGoOqW+OsH5PybqoN+uMxKgC-soIEVZbA0iwVOOeyNF4KRXlg9erJN68h-u3P+ABmQBwDQGD2WCPIMEwT4GmUFAmBskNhehvogye3A64z0AurUhGCCoVxwUoOhHcAAszDD5xzYcPK+MJuGQIHPwkEgi4RfC0Vw8eIi75gBQRIueMiP7wBboov+ABWVRfd1HcP0aCbR49REGDRNAQQpwpTXjpIKK4oBHR7zGHIQYWCJJSWwnACwaxTLfh4ROACM4xi8hhurECgwWRkPfrYSOm5o69SAA)

**Comment by Convly** [comment]
Well, so the issue here boils down to TypeScript's conditional and distributive behavior with union types. The union type (`TSchemaUID`) gets distributed across the types within `ObjectNotation`.

Instead of returning
```ts
AttributesFiltering<'api::a.a'> | AttributesFiltering<'api::b.b'> | ...
```
it's doing
```ts
AttributesFiltering<'api::a.a' | 'api::b.b' | ... >
```

under the hood

This distributive property of generics means that TypeScript has to keep every possibility open when building `ObjectNotation` (and thus `AttributesFiltering`), which essentially means that it leaves any `Common.UID.Schema` as open possibilities rather than narrowing it to one.

It seems that the reason it doesn't break when using non (literal) union types is because TypeScript needs more context or stricter type constraints to infer the exact type from the union. `enumeration` attributes' values are not considered as one type but as a union of multiple types and TS cannot guarantee which one is being inferred.

I managed to fix the faulty behavior by creating an intermediary mapping step for `TSchemaUID` in `ObjectNotation`. This allows TypeScript's generic inference to correctly distribute the `TSchemaUID` union into the individual keys of `AttributesFiltering` and `RootLevelOperatorFiltering` types.

In short, going from

```ts
export type ObjectNotation<TSchemaUID extends Common.UID.Schema> = RootLevelOperatorFiltering<TSchemaUID> & AttributesFiltering<TSchemaUID>;
```

to

```ts
export type ObjectNotation<TSchemaUID extends Common.UID.Schema> =
  TSchemaUID extends infer TUIDs extends Common.UID.Schema
    ? // The intermediary mapping step below allows TypeScript's generic inference to correctly distribute the
      // TSchemaUID union into the individual keys of AttributesFiltering and RootLevelOperatorFiltering types
      {
        [TUID in TUIDs]: RootLevelOperatorFiltering<TUID> & AttributesFiltering<TUID>;
      }[TSchemaUID]
    : never;
```

---

On a side note, now that we've implemented this intermediate step in `ObjectNotation`, we don't need any additional conditional statements within `AttributesFiltering` members.
Indeed, with the modification above, TypeScript will now correctly distribute the `TSchemaUID`. This essentially makes TypeScript check each key of the union individually ensuring that TS can infer the types correctly down the line.

I haven't removed the conditionals from there though, as one could use those types directly without using the same distribution strategy as stated in `ObjectNotation`.

File: packages/core/types/src/modules/entity-service/params/filters/index.ts (hunk 3 above)
Comment (Convly) [reply packages/core/types/src/modules/entity-service/params/filters/index.ts:66]: We might consider that later, for now I'll keep the same behavior as before regarding the ID filters and just fix the issue at hand (trying to avoid adding side effect to the PR)

Comment (Convly) [reply packages/core/types/src/modules/entity-service/params/filters/index.ts:66]: Thanks for the suggestion though

**Review by innerdvations — APPROVED** [review]
YOU GOT IT!

This looks good, makes sense, and I did some QA on the cases we found (multiple sorts+multiple filters in the same query) that were broken before, and they work now.
